bin/propolis-server/src/lib/vm/state_driver.rs

```diff
@@ -520,20 +534,167 @@ impl StateDriver {
-                                 "error" => ?e);
-                self.publish_steady_state(InstanceState::Failed);
+        }
+
+        // Next, prepare to start block backends. This is done by capturing the
+        // current block backend set and creating a future that issues all the
+        // start requests.
+        //
+        // For this to work, the set of block backends to be started must not
+        // change while the VM is starting. This is guaranteed because all such
+        // requests to hotplug a block backend will be dispatched to the VM's
+        // request queue; if any such requests are seen below, they can simply
+        // be buffered and handled after the rest of the VM has started.
+        async fn start_block_backends(
+            log: slog::Logger,
+            backends: BlockBackendMap,
+        ) -> anyhow::Result<()> {
+            for (name, backend) in backends {
+                info!(log, "starting block backend {}", name);
+                let res = backend.start().await;
+                if let Err(e) = &res {
+                    error!(
+                        log,
+                        "startup failed for block backend {}: {:?}", name, e
+                    );
+
+                    return res;
+                }
             }
+
+            Ok(())
         }
 
-        start_result
+        let block_backends = objects.block_backend_map().clone();
+        let block_backend_fut =
+            start_block_backends(self.log.clone(), block_backends);
+        tokio::pin!(block_backend_fut);
+
+        // Drop the VM object lock before proceeding to allow other API calls
+        // that simply read the VM to make progress. Again, note that the set of
+        // objects being started still can't change, not because the lock is
+        // held, but because the only entity that can change them is the current
+        // task, which can decide whether and how to buffer incoming requests.
+        drop(objects);
+
+        loop {
+            let event = tokio::select! {
+                // If the VM successfully starts, return immediately and let
+                // the caller process any events that may happen to be on the
+                // queue.
+                biased;
+
+                res = &mut block_backend_fut => {
+                    if res.is_ok() {
+                        let objects = &self.objects;
+                        objects.lock_exclusive().await.resume_vcpus();
+                        self.publish_steady_state(InstanceState::Running);
+                        info!(&self.log, "VM successfully started");
+                    }
+
+                    return res;
+                }
+
+                dequeued = self.input_queue.wait_for_next_event() => {
+                    dequeued
+                }
+            };
+
+            // The VM's vCPUs haven't been started yet, so there should be no
+            // way for the VM to produce any guest events to handle.
```

Comment:
Is there a possibility for a race condition here?  I don't think so, because there are no yield points between starting the vCPUs and returning from the `block_backend_fut`, but want to double-check (and maybe suggest a comment to that effect).

Reply:
Added a comment in bfb8b07: IIUC, this shouldn't race because vCPUs only start when the `block_backend_fut` branch of the `select!` is actually chosen, and once that happens the whole function returns, so if we're here then we haven't had an occasion to start the vCPUs yet.